### PR TITLE
Add Common Subexpression Elimination

### PR DESCRIPTION
**Pull request by jvican**
More information in the commit messages and source code.

### CODE REVIEW

**Comment by jvican** [comment]
Review by @DarkDimius.

File: src/dotty/tools/dotc/transform/ElimCommonSubexpression.scala [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+  *
+  * will be transformed to:
+  * {{{
+  *   val a = 1
+  *   val a1 = a + a
+  *   val b = a1 + 2
+  *   val c = a1 + 2
+  * }}}
+  *
+  * only if `+` is guaranteed to be immutable.
```
Comment (jvican) [inline src/dotty/tools/dotc/transform/ElimCommonSubexpression.scala:39]: s/immutable/idempotent/g

**Comment by jvican** [comment]
I've added a final version of the transformation that works transparently for more test cases. Before, my optimization wasn't able to apply the optimization in concrete cases and the logic was rough. As a result, a lot of cases were not handled.

Another important note is that now the optimization interacts with itself, and it's able to reuse already applied optimizations of idempotent sub trees. See `test5`.

I'd like to know if I should add a proper test infrastructure, right now i'm executing the test files by hand.

For more information about my changes, see the commit message:
- This final version is able to detect common subexpression in rhs that
  are impure but contain idempotent trees. It also makes it work more
  transparently, and removes some bad assumptions and rough edges of the
  previous implementation.
- Add more complex test cases to show functionality.
- Change basic algorithm and make it more performant.
- Significantly simplify the way the algorithm works (in the logical
  level).

**Comment by DarkDimius** [comment]
> I'd like to know if I should add a proper test infrastructure, right now i'm executing the test files by hand.

We would work on rebuilding test infrastructure after rebasing over latest HK-types in Dotty.

**Comment by allanrenucci** [comment]
I have some tests you might want to reuse: https://github.com/allanrenucci/dotty/blob/idempotent-call-elimination/tests/run/idempotentcalls.scala

**Comment by jvican** [comment]
Thanks, I’m basically doing my own test suite exactly like yours, and it’s almost finished (although mine is more compact and short). I’ll have a look before shipping this to make sure all of your tests pass as well :wink:.

**Comment by jvican** [comment]
This PR is completely outdated, I'm closing it. The new optimization is implemented in the branch `cse-rebase`.

**Comment by jvican** [comment]
https://github.com/dotty-linker/dotty/tree/cse-rebase

**Pull request closed without merging**